File: Problems/Previous Problems/Connect ropes/15-heap-connect-ropes/code.cpp
#include<bits/stdc++.h> 
using namespace std; 

int ConnectRopes(vector<int>length, int n) 
{ 
	int sum = 0; 
	priority_queue<int, vector<int>, greater<int> > pq; 

	for(int i = 0; i < n; i++) 
		pq.push(length[i]); 

	while (pq.size() > 1) 
	{ 
		int min = pq.top(); 
		pq.pop(); 
		int secondMin = pq.top(); 
		pq.pop(); 
		sum += (min + secondMin); 
		pq.push(min + secondMin); 
	} 
	return sum; 
} 

int main() 
{	
	int t;
    cin>>t;
    while(t--)
    {
    	int n;
    	cin>>n;
    	vector<int> length(n);
    	for (int i = 0; i < n; ++i)
    		cin>>length[i];
    	cout << (connectRopes(length, n))<<"\n"; 
    }
    return 0;
} 

File: Problems/Previous Problems/Connect ropes/15-heap-connect-ropes/code.py

import heapq 
def ConnectRopes(length, n): 
	sum = 0
	pq = length
	heapq.heapify(pq) 
	
	while (len(pq) > 1) : 
		min = pq[0]; 
		heapq.heappop(pq); 
		secondMin = pq[0]; 
		heapq.heappop(pq); 
		sum += (min + secondMin); 
		heapq.heappush(pq , min + secondMin) 
	return sum


if __name__ == '__main__': 
    t = int(input())
    for _ in range(t):
        n = int(input())
        length = list(map(int, input().strip().split()))
        print(ConnectRopes(length,n))


